lib/ueberauth/strategy.ex

```diff
@@ -297,17 +316,16 @@ defmodule Ueberauth.Strategy do
 
   @doc false
   def run_request(conn, strategy) do
-    apply(strategy, :handle_request!, [conn])
+    conn
+    |> maybe_add_state_param(strategy)
+    |> run_handle_request(strategy)
   end
 
   @doc false
   def run_callback(conn, strategy) do
-    handled_conn =
-      strategy
-      |> apply(:handle_callback!, [conn])
-      |> handle_callback_result(strategy)
-
-    apply(strategy, :handle_cleanup!, [handled_conn])
+    strategy
+    |> get_ignores_csrf_attack_option()
```

Comment:
Could we not add a if statement here, I think the current does not ready very easily.

Basically we have two paths
If `csrf_protection` is enabled
 if `state_param_matches` succeseds
 then `run_handle_callback`
 else run `add_state_mismatch_error`
else `run_handle_callback`

Or maybe move this logic to its own module?

Reply:
@Hanspagh I am sorry, I am not following what I suppose to do

Reply:
For me the pipeline gets really hard to follow with all these small functions with control flow inside them.

My suggestion was ether to make a dedicated csrf_protection module, where we can do the cookie checking logic.

Or

To avoid having the control flow in the functions by using if

Reply:
@yordis I think @Hanspagh is suggesting something like this:

```elixir
def run_callback(conn, strategy) do
  if get_ignores_csrf_attack_option(strategy) do
    if state_param_matches?(conn) do
      run_handle_callback(conn, strategy)
    else
      add_state_mismatch_error(conn, strategy)
    end
    run_handle_callback(conn, strategy)
  end
end
```